src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_ecdsa.c

```diff
@@ -0,0 +1,112 @@
+    (*env)->DeleteLocalRef(env, privateKey);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+
+    jbyteArray digestArray = (*env)->NewByteArray(env, dgstlen);
+    (*env)->SetByteArrayRegion(env, digestArray, 0, dgstlen, (const jbyte*)dgst);
+    (*env)->CallVoidMethod(env, signatureObject, g_SignatureUpdate, digestArray);
+    (*env)->DeleteLocalRef(env, digestArray);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+
+    jbyteArray sigResult = (*env)->CallObjectMethod(env, signatureObject, g_SignatureSign);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+    jsize sigSize = (*env)->GetArrayLength(sigResult);
+    (*env)->GetByteArrayRegion(env, sigResult, 0, sigSize, sig);
+    (*env)->DeleteLocalRef(env, sigResult);
+
+    (*env)->DeleteLocalRef(env, signatureObject);
+
+    return SUCCESS;
+
+    error:
+    (*env)->DeleteLocalRef(env, signatureObject);
+    return FAIL;
+}
+
+int32_t
+CryptoNative_EcDsaVerify(const uint8_t* dgst, int32_t dgstlen, const uint8_t* sig, int32_t siglen, EC_KEY* key)
+{
+    JNIEnv* env = GetJNIEnv();
+
+    jobject signatureObject = CryptoNative_GetEsDsaSignatureObject();
+    if (!signatureObject)
+    {
+        return FAIL;
+    }
+
+    jobject publicKey = (*env)->CallObjectMethod(env, key->keyPair, g_keyPairGetPublicMethod);
+    (*env)->CallVoidMethod(env, signatureObject, g_SignatureInitVerify, publicKey);
+    (*env)->DeleteLocalRef(env, publicKey);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+
+    jbyteArray digestArray = (*env)->NewByteArray(env, dgstlen);
+    (*env)->SetByteArrayRegion(env, digestArray, 0, dgstlen, (const jbyte*)dgst);
+    (*env)->CallVoidMethod(env, signatureObject, g_SignatureUpdate, digestArray);
+    (*env)->DeleteLocalRef(env, digestArray);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+
+    jbyteArray sigArray = (*env)->NewByteArray(env, dgstlen);
+    (*env)->SetByteArrayRegion(env, sigArray, 0, siglen, (const jbyte*)sig);
+    (*env)->CallObjectMethod(env, signatureObject, g_SignatureVerify, sigArray);
+    (*env)->DeleteLocalRef(env, sigArray);
+    ON_EXCEPTION_PRINT_AND_GOTO(error);
+
+    (*env)->DeleteLocalRef(env, signatureObject);
+
+    return SUCCESS;
+
+    error:
+    (*env)->DeleteLocalRef(env, signatureObject);
+    return FAIL;
+}
+
+int32_t CryptoNative_EcDsaSize(const EC_KEY* key)
+{
+    // The maximum size of a signature for the provided key is 2* bitlength of the order + extra bytes for the DER encoding.
+    // The DER encoding is as follows (with R and S being the components of the signature and all lengths being one byte width):
+    // - SEQUENCE <length of sequence> INTEGER <length of R> <R> INTEGER <length of S> <S>
+    // As a result, we see that there are 6 additional bytes in the DER encoding than the lengths of R and S combined.
+    // As the ECDSA algorithm is defined, the maximum length of R and S each is the bitlength of the order, so as a result
+    // we get the maximum size as 2 * bitlength of the order + 6.
+    const int derEncodingBytes = 6;
```

Comment:
The DER encoding overhead is not a constant.

Each of R and S get wrapped in an integer, which has a 1 byte tag (02) and a variable-byte length (1 for < 0x80, 2 for 0x80-0xFF, 3 for 0xFF-0xFFFF, 4 for 0xFFFF-0xFFFFFF, ...), then those two values get wrapped in a SEQUENCE (1 byte tag, same variable length).  Integers with the high bit set will gain a padding byte to prevent them from being treated as a negative number.

So for secp256r1 the maximum length for the integer contents is `1 + 256/8` = 33 bytes.  33 bytes produce a prefix of 0221.  The maximum given to the sequence is 2x35 bytes == 70 bytes, so it'll start with (maximally) 3046.  6 bytes held.

For secp521r1 a surprise padding byte is never needed (since 521 is not byte aligned).  It needs a maximum of (521+7)/8 bytes (66) for R or S's content, so the prefix is 0242.  But, put two of those together, and you have 68*2 = 136 bytes for the sequence.  That's more than 127, so the prefix is 308188... 3 bytes.  This one took 7.

So 7 is the practical maximum (there aren't really bigger curves yet, because they're 'pointlessly secure').

If you haven't had an assert trip on this yet it's because it requires both R and S to have their highest bit set, which sounds like 25%, but it's less than that since there are fewer legal values with the bit set than without them being set.  I feel like it should come out to around 18% of the time that it needed 7 bytes of overhead, but that's recall off the top of my head, not using math freshly.

Reply:
Java doesn't supply a method to get the size since they provide an API that returns an array of the right length, so we have to handle the calculation ourselves. Maybe we can calculate the max DER overhead based on the length of the order + the extra bytes for the padding bytes?

Reply:
Can you use the "inP1363Format" versions described in https://docs.oracle.com/en/java/javase/12/docs/specs/security/standard-names.html#signature-algorithms?  Their documentation is wrong, but hopefully their code is right.

The P1363 format is the .NET default, r and s as a fixed-length unsigned encoding each as long as the size of the order.

Of course, that would mean you have to split off from ECDsaOpenSsl already, since ECDsaOpenSsl expects the X.509 DER version.

Reply:
Yes, I should be able to do that. I was hoping to avoid having to make changes to managed code in this PR, but I can do it if need be.

Reply:
Android doesn't support the "inP1363Format" variations, so it looks like we'll have to fix this.

Reply:
The _easiest_ fix is to set the overhead to 18.  That would be where each of R and S are between 0x0100_0000 and 0x7FFF_FFFE long and the resulting sequence then contains 0xFFFF_FFFE payload bytes (each segment adds a 1 byte tag and 5 byte length, 6*3 = 18).

The hybrid of easy and not wasteful would be to set the overhead to 7, which covers up through secp521r1.  (It's good for until R/S exceed 0xFF bytes, which would be at 255*8-1 (2039) bits... so probably good enough).

For perfect accuracy, you'll want something like https://source.dot.net/#System.Formats.Asn1/System/Formats/Asn1/AsnWriter.cs,aa1dbe1cb6c5929e (though you need to add one for the "non-subsequent" byte that this class already knew about, and one for the tag), and to do 

```
int rsMax = SegmentSize((keySizeInBits + 8) / 8);
int seqMax = SegmentSize(rsMax * 2);
return seqMax;
```

Why +8 / 8?  "Add one bit for the sign bit, then it's bytes, round up".  0-7 bits: 1 byte.  8-15: 2 bytes, etc.

But, that'll just return 6 for KeySize <= (((0x7F / 2) - 2) * 8) - 1 = 487 bits, and 7 for up to 2039 bits.  The biggest defined in OpenSSL is sect571r1, at 571 bits... so 7 is probably safe.

Reply:
I've set it to 7 for now.